File: src/main/java/main/struct/CompositePart.java
package main.struct;


import java.util.HashSet;
import java.util.Set;

public class CompositePart extends designObject {
    public String type = "cp";

    Set<BaseAssembly> usedIn = new HashSet<>();
    Set<AtomicPart> atomicParts = new HashSet<>();
    AtomicPart rootPart;

    public CompositePart(long id, int buildDate) {
        super(id, buildDate);
    }

    public boolean addAssembly(BaseAssembly b){
        return usedIn.add(b);
    }

    public boolean removeAssembly(BaseAssembly b){
        return usedIn.remove(b);
    }

    public boolean addPart(AtomicPart a){
        return atomicParts.add(a);
    }

    public boolean removePart(AtomicPart a){
        return atomicParts.remove(a);
    }

    public Set<AtomicPart> getAtomicParts() {
        return atomicParts;
    }

    public Set<BaseAssembly> getUsedIn() {
        return usedIn;
    }
}

File: src/struct/composite_part.cc
#include "atomic_part.h"
#include "composite_part.h"
#include "document.h"

using namespace std;

bool sb7::CompositePart::addPart(AtomicPart *apart) {
    bool add = m_parts->add(apart);

    if (add) {
        apart->setCompositePart(this);

        if (m_rootPart == NULL) {
            m_rootPart = apart;
        }
    }

    return add;
}
